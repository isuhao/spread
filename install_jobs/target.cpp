#include "target.hpp"

#include <htasks/copyhash.hpp>
#include <htasks/downloadhash.hpp>
#include <htasks/unpackhash.hpp>
#include <stdexcept>

using namespace Spread;

std::string Target::fetchFile(const Hash &hash)
{
  JobPtr fetchTask;
  std::string output = owner->fetchFile(hash, fetchTask);
  assert(output != "");

  /* If there is a job required to fetch this file, then fetchFile()
     will have set it up for us. Execute it.
  */
  if(fetchTask)
    execJob(fetchTask, true);

  return output;
}

bool Target::execHashTask(HashTask *htask, bool failOnError)
{
  assert(htask);
  assert(!src.hash.isNull());

  // Add inputs
  for(int i=0; i<src.deps.size(); i++)
    {
      const Hash &input = src.deps[i];
      assert(!input.isNull());
      assert(input != src.hash);
      htask->addInput(input, fetchFile(input));
    }

  // Add outputs
  Hash::DirMap::const_iterator it;
  for(it = output.begin(); it != output.end(); it++)
    htask->addOutput(it->second, it->first);

  return execJob(htask, failOnError);
}

void Target::doJob()
{
  assert(owner);
  if(src.type == TST_File)
    execHashTask(new CopyHash(src.value), true);
  else if(src.type == TST_Download)
    {
      if(!execHashTask(new DownloadHash(src.value), false))
        {
          // Let the caller try to find a replacement
          owner->brokenURL(src.hash, src.value);
          throw std::runtime_error(lastJob->getMessage());
        }
    }
  else if(src.type == TST_Archive)
    {
      assert(src.dir);
      execHashTask(new UnpackHash(*src.dir), true);
    }
  else assert(0);
  if(checkStatus()) return;

  /* The order is important here: we might have other threads waiting
     for us, thus we cannot set 'done' status until AFTER have added
     our output files to cache.

     notifyFiles() implementations must also make sure to add the
     files to cache before removing them from target lists used by
     other jobs.
   */
  assert(output.size());
  owner->notifyFiles(output);
  setDone();
}
