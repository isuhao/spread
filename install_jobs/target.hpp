#ifndef __SPREAD_TARGET_HPP_
#define __SPREAD_TARGET_HPP_

#include <install_system/ihashfinder.hpp>
#include <parent_job/execjob.hpp>

namespace Spread
{
  struct TargetOwner
  {
    virtual std::string fetchFile(const Hash &hash, JobPtr &job,
                                  const std::string &target="") = 0;
    virtual void brokenURL(const Hash &hash, const std::string &url) = 0;
    virtual void notifyFiles(const Hash::DirMap &files) = 0;
  };

  struct HashTask;
  struct Target : ExecJob
  {
    HashSource src;
    Hash::DirMap output;
    Target(TargetOwner *o) : owner(o) {}

  private:
    TargetOwner *owner;
    std::string fetchFile(const Hash &hash);
    bool execHashTask(HashTask *htask, bool failOnError);
    void doJob();
  };

  typedef boost::shared_ptr<Target> TargetPtr;
}
#endif
