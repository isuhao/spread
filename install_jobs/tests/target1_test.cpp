#include "target.hpp"
#include <boost/filesystem.hpp>
#include <iostream>

using namespace std;
using namespace Spread;
namespace bf = boost::filesystem;

Hash hello("hello", 5);
Hash world("world", 5);
Hash robots("N0VT8AYfLEu2hFufocgj9ykAQoNEgcQwzLW7m1Tfc-cj");
Hash testsh("ctEjJBRstghw4_UpmjBdhwJZFl8faISyIeEk2sOH5LLfAQ");
Hash dolly("rSdU-hHwettk1icc_gOrTKGJKe3BWeMSCHFkDKgmnf4M");
Hash zipH("UZuuyrHbX1c57drq6a6SObKZQrkr58Y09SrBFtEHm9rSAg");

struct TestOwner : TargetOwner
{
  std::string fetchFile(const Hash &hash, JobPtr &job, const std::string &target)
  {
    assert(target == "");
    cout << "fetchFile(" << hash << ")\n";
    if(hash == zipH) return "test.zip";
    assert(0);
  }

  void brokenURL(const Hash &hash, const std::string &url)
  {
    cout << "Reporting broken URL=" << url << " HASH=" << hash << endl;
  }

  void notifyFiles(const Hash::DirMap &files)
  {
    cout << "Adding " << files.size() << " files to cache:\n";
    Hash::DirMap::const_iterator it;
    for(it = files.begin(); it != files.end(); it++)
    {
      const std::string &file = it->first;
      const Hash &hash = it->second;
      cout << "  " << hash << " " << file << endl;
    }
  }
};

TestOwner owner;

void run(Job &j)
{
  cout << endl;
  JobInfoPtr info = j.run();
  cout << "  msg=" << info->getMessage()
       << "\n  prog=" << info->getCurrent() << "/" << info->getTotal()
       << endl;
  if(info->isSuccess()) cout << "  SUCCESS!\n";
  else if(info->isError()) cout << "  FAILURE!\n";
  else assert(0);
}

int main()
{
  bf::remove_all("_target");
  {
    Target t(&owner);
    t.src.type = TST_File;
    t.src.value = "hello.dat";
    t.src.hash = hello;
    t.output["_target/hello"] = hello;
    run(t);
  }
  {
    Target t(&owner);
    t.src.type = TST_File;
    t.src.value = "hello.dat";
    t.src.hash = hello;
    t.output["_target/hello_fail"] = world;
    run(t);
  }
  {
    Target t(&owner);
    t.src.type = TST_File;
    t.src.value = "world.dat";
    t.src.hash = world;
    t.output["_target/world1"] = world;
    t.output["_target/world2"] = world;
    run(t);
  }
  /* Uncomment to test downloads
  {
    Target t(&owner);
    t.src.type = TST_Download;
    t.src.value = "http://tiggit.net/robots.txt";
    t.src.hash = robots;
    t.output["_target/robo1"] = robots;
    run(t);
  }
  //*/
  {
    Target t(&owner);
    t.src.type = TST_Download;
    t.src.value = "http://doesnt/exist";
    t.src.hash = robots;
    t.output["_target/robo2"] = robots;
    run(t);
  }

  Hash::DirMap dir;
  dir["test.sh"] = testsh;
  dir["dir/dolly.txt"] = dolly;

  {
    Target t(&owner);
    t.src.type = TST_Archive;
    t.src.hash = testsh;
    t.src.dir = &dir;
    t.src.deps.push_back(zipH);
    t.output["_target/test"] = testsh;
    run(t);
  }

  return 0;
}
