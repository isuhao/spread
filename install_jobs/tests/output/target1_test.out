
Adding 1 files to cache:
  LPJNul-wow4m6DsqxbninhsWHlwfp0JecwQzYpOLmCQF _target/hello
  msg=Copying hello.dat
  prog=5/5
  SUCCESS!

  msg=Hash mismatch in _target/hello_fail
  Expected: SG6kYiTRu0-2gPNPfJrZao8k7Ii-c-qOWmxlJg6cuKcF
  Recieved: LPJNul-wow4m6DsqxbninhsWHlwfp0JecwQzYpOLmCQF
  prog=5/5
  FAILURE!

Adding 2 files to cache:
  SG6kYiTRu0-2gPNPfJrZao8k7Ii-c-qOWmxlJg6cuKcF _target/world1
  SG6kYiTRu0-2gPNPfJrZao8k7Ii-c-qOWmxlJg6cuKcF _target/world2
  msg=Copying world.dat
  prog=5/5
  SUCCESS!

Reporting broken URL=http://doesnt/exist HASH=N0VT8AYfLEu2hFufocgj9ykAQoNEgcQwzLW7m1Tfc-cj
  msg=Error fetching http://doesnt/exist:
Couldn't resolve host name
  prog=0/0
  FAILURE!

fetchFile(UZuuyrHbX1c57drq6a6SObKZQrkr58Y09SrBFtEHm9rSAg)
Adding 1 files to cache:
  ctEjJBRstghw4_UpmjBdhwJZFl8faISyIeEk2sOH5LLfAQ _target/test
  msg=Unpacking test.zip
  prog=479/479
  SUCCESS!
